org.osgi.test.common/src/main/java/org/osgi/test/common/dictionary/Dictionaries.java

```diff
@@ -441,6 +443,83 @@ public String toString() {
 			.collect(toDictionary());
 	}
 
+	private static class ServiceReferenceAsDictionary extends Dictionary<String, Object> {
+		private final ServiceReference<?> serviceReference;
+
+		@SuppressWarnings("unchecked")
+		ServiceReferenceAsDictionary(ServiceReference<?> serviceReference) {
+			this.serviceReference = requireNonNull(serviceReference);
+			boolean nullKey = keysInternal().contains(null);
+
+			if (nullKey) {
+				throw new NullPointerException("a Dictionary cannot contain a null key");
+			}
+			boolean nullValue = false;
+
+			for (String key : keysInternal()) {
+				if (Objects.isNull(serviceReference.getProperty(key))) {
+					nullValue = true;
+					break;
+				}
+			}
+
+			if (nullValue) {
+				throw new NullPointerException("a Dictionary cannot contain a null value");
+			}
+		}
+
+		private List<String> keysInternal() {
+			String[] keys = serviceReference.getPropertyKeys();
+			return Objects.isNull(keys) ? Collections.EMPTY_LIST : Arrays.asList(keys);
+		}
+
+		@Override
+		public int size() {
+			return keysInternal().size();
+		}
+
+		@Override
+		public boolean isEmpty() {
+			return keysInternal().isEmpty();
+		}
+
+		@Override
+		public Enumeration<String> keys() {
+			return Collections.enumeration(keysInternal());
+		}
+
+		@Override
+		public Enumeration<Object> elements() {
+			List<Object> values = new ArrayList<>();
+			for (String key : keysInternal()) {
+				values.add(serviceReference.getProperty(key));
+			}
+			return Collections.enumeration(values);
+		}
+
+		@Override
+		public Object get(Object key) {
+			if (key instanceof String || key == null) {
+				return serviceReference.getProperty((String) key);
+			}
+			throw new IllegalArgumentException();
+		}
+
+		@Override
+		public Object put(String key, Object value) {
+			throw new UnsupportedOperationException();
+		}
+
+		@Override
+		public Object remove(Object key) {
+			throw new UnsupportedOperationException();
+		}
+	}
+
+	public static <K, V> Dictionary<String, Object> asDictionary(ServiceReference<?> serviceReference) {
+		return new ServiceReferenceAsDictionary(serviceReference);
```

Comment:
Why do we need this? ServiceReference has a getProperties method which returns a dictionary which can then be asserted on. https://docs.osgi.org/javadoc/osgi.core/7.0.0/org/osgi/framework/ServiceReference.html#getProperties--

Reply:
Oh, this came with R7. `osgi-test` uses R6 for core. Could we update to R7?

The goal is to get a fast and fluent way to came from `ServiceEventAssert` over `ServiceReferenceAssert` to `DictionaryAssert` via `has...That()`

Reply:
Agreed, it would be simpler to use ServiceReference.getProperties(), but as @stbischof pointed out it's not available until R7.

Reply:
So perhaps we update the osgi.core dependency to 7.0.0? @rotty3000 ?

Reply:
For posterity: it was resolved to keep R6 compatibility so `ServiceReferenceAsDictionary` was retained.